package eos;

message GetTimestamps {
}

message Timestamps {
  required uint64 base_map = 1;
  required uint64 static_map = 2;
  required uint64 display_map = 3;
  required uint64 patrols = 4;
  required uint64 zones = 5;
  required uint64 sensors = 6;
}

message Zone {
  enum Permission {
    Nothing = 0;
    NoDetection = 1;
    NoMovement = 2;
    NoRemapping = 3;
    Map = 4;
  }
  message Point {
    required uint32 x = 1;
    required uint32 y = 2;
  }
  required Point top_left = 1;
  required Point bottom_right = 2;
  required Permission permissions = 3 [default = NoDetection];
}

message Zones {
  required uint64 timestamp = 1 [default = 0];
  repeated Zone zones = 2;
}

message JSONSettings {
 required string keys = 1;
 required string value = 2;
}

message Settings {
  required uint64 timestamp = 1 [default = 0];
  repeated JSONSettings settings = 2;
}

message Range {
  required uint32 min = 1 [default = 0];
  required uint32 max = 2 [default = 0];
}

message Sensor {
  message Point {
    required uint32 x = 1 [default = 0];
    required uint32 y = 2 [default = 0];
  }
  enum Type {
    RobotDock = 0;
    DoorSwitch = 1;
    Radar = 2;
    InfraredBarrier = 3;
    Rack = 4;
  }
  required string name = 1;
  required Point position = 2;
  required Point position_second_element = 3;
  required int32 orientation = 4 [default = 0];
  required Range distance = 5;
  required Range angle = 6;
  required Type type = 7 [default = RobotDock];
  required Point robot_point = 8;
}

message Sensors {
  required uint64 timestamp = 1 [default = 0];
  repeated Sensor sensors = 2;
}

message Patrol {
  enum Type {
    Normal = 0;
    PopOut = 1;
    Remapping = 2;
    Refill = 3;
  }

  enum Priority {
    Standard = 0;
    High = 1;
  }

  message Position {
    required uint32 x = 1;
    required uint32 y = 2;
  }

  message Point {
    required Position position = 1;
    required Priority priority = 2;
  }

  required string name = 1;
  required Type type = 2 [default = Normal];
  repeated Point points = 3;
  //repeated Priority priorities = 4 ;
}

message Patrols {
  required uint64 timestamp = 1 [default = 0];
  repeated Patrol Patrols = 2;
}

message Map {
  enum Type {
    BASE_MAP = 0;
    REMAP_MAP = 1;
    DISPLAY_MAP = 2;
  }
  required uint64 timestamp = 1 [default = 0];
  required uint32 width = 2;
  required uint32 height = 3;
  required float resolution = 4;
  required bytes content = 5;
  required Type type = 6 [default = BASE_MAP];
}

message TaskScheduled {
  required string patrol_name = 1 [default = "none"];
  required int64 start = 2 [default = 0];
  required int64 stop = 3 [default = 0];
  required int64 group_id = 4 [default = 0];
}

message TasksScheduled {
  required uint64 timestamp = 1 [default = 0];
  repeated TaskScheduled tasks = 2;
}

